### ISSUE TITLE
props and options get called multiple times while fetching a query

### ISSUE THREAD

**Issue by ablbol** [apollographql/react-apollo#660]
I am using react apollo version "^1.0.0" and trying to fetch a query.  While debugging my code I noticed that the "options" and "props" parts called several times before the query is fetched (see comments in the code snippet).  This behavior happens in other queries in my code.  Is there a reason for it or did I miss something.  Here is the container where I am fetching a query

```
export default graphql(query, {
  name: 'users',

  // this function is called several times by Apollo client per query
  props: ({ users: { loading, users, fetchMore } }) => ({........}),

  // this function is called several times by Apollo client per query
  options: (props) => ( {variables: {......}} ) 

)(Users);

```

**Comment by helfer** [comment]
Hi @ablbol. I'm not too familiar with that part of the codebase, but if you provide a reproduction with [react-apollo-error-template](https://github.com/apollographql/react-apollo-error-template), I'll look into it!

Or maybe @jbaxleyiii has an idea! :)

**Comment by jbaxleyiii** [comment]
A reproduction would be great! If I had to hazard a guess I would say that something else is rerendering your wrapped component while the query is in flight. Options and props get called each time the component is rerendered if the props coming down the tree have changed

**Comment by newoga** [comment]
@helfer @jbaxleyiii 

I know @ablbol closed this issue, however I'm seeing this behavior particularly with the `options()` method, not the `props()` method.

A very simple reproduction can be found here: https://github.com/newoga/react-apollo-issue-660

This is the console output from the reproduction when the application starts:

![image](https://user-images.githubusercontent.com/344018/27511431-cff13986-58f1-11e7-9b24-86cdfa0bc14a.png)

**Comment by newoga** [comment]
FYI I did a little digging myself. I added additional logging in my `node_modules\react-apollo\react-apollo.browser.umd.js` one line before each call of `calculateOptions` (which calls `mapPropsToOptions`) and each direct call of  `mapPropsToOptions`.

It seems like there are a number of methods that each recalculates the options for themselves for the same props.

Below is the log output I have. It shows each `react-apollo` client method that is calling the user's option mapping function.

```
[node_modules/react-apollo] shouldSkip() -- mapPropsToOptions
App.js:49 options() called...

App.js:50 
react-apollo.browser.umd.js:286 [node_modules/react-apollo] setInitialProps -- calculateOptions
App.js:49 options() called...

App.js:50 
react-apollo.browser.umd.js:381 [node_modules/react-apollo] shouldSkip() -- mapPropsToOptions
App.js:49 options() called...

App.js:50 
react-apollo.browser.umd.js:406 [node_modules/react-apollo] dataForChild -- calculateOptions
App.js:49 options() called...

App.js:50 
App.js:6 App rendering...

react-apollo.browser.umd.js:381 [node_modules/react-apollo] shouldSkip() -- mapPropsToOptions
App.js:49 options() called...

App.js:50 
react-apollo.browser.umd.js:381 [node_modules/react-apollo] shouldSkip() -- mapPropsToOptions
App.js:49 options() called...

App.js:50 
react-apollo.browser.umd.js:406 [node_modules/react-apollo] dataForChild -- calculateOptions
App.js:49 options() called...

App.js:50 
App.js:6 App rendering...

```

I haven't investigate the source code enough to provide any recommendations to circumvent this, but hopefully this adds some clarity to the problem.

**Comment by newoga** [comment]
@helfer @jbaxleyiii I haven't had a chance to explore this any further, but do you think we should reopen this issue? If I get around to it, I can try to write some failing tests for this.

**Comment by Mati365** [comment]
Any updates?

**Comment by stubailo** [comment]
Reopening since there's now a reproduction!

**Comment by sbr464** [comment]
I noticed this same problem, A component that take a variable / id from the router params, will re-render 7-8 times each time the component loads.


On a full page refresh It reloads 8 total times -  5 times, then when the data comes back from the graphql endpoint, it runs 3 more times. The difference in the 2 objects are that `loading` is false, and `account`  has data once loaded.

<img width="792" alt="screen shot 2017-09-17 at 12 12 26 pm" src="https://user-images.githubusercontent.com/8600215/30523605-85117000-9ba1-11e7-9111-857a4cbf336f.png">

**Comment by Mati365** [comment]
@jbaxleyiii can you reproduce this problem?

**Comment by akaipham** [comment]
Any updates guys?
```javascript
const HomeScreenWithData = graphql(query, {
  options: ({currentUser}) => {
    const _ownerId = currentUser ? currentUser._id : '';
    return {variables: {_ownerId}};
  },
  props: props => props,
})(HomeScreen);
```
Here's my simple code to fetch user data but it seems fetching 4 times then trigger rendering 4 times as well. :\

**Comment by jmlivingston** [comment]
I'm seeing this behavior as well. I removed all apollo code to see if it was something router related, but that was not the problem. The one good thing I noticed was my GraphQL server wasn't getting hit multiple times, only once. It's just the options function that gets called multiple times.

**Comment by acomito** [comment]
Same here. Seeing it call about 7-8 times in the client (added a console.log to my function that return options for graphql()() )

**Comment by cmccand** [comment]
@jbaxleyiii I am seeing this as well, and it's really making pagination difficult

**Comment by rchasman** [comment]
Same here. I'm wondering if I pass a route that doesn't match an ID, if there is a way to abort the query. Right now I am getting infinte retries when the ID doesn't match, which asplodes the browser.

**Comment by thiago-soliveira** [comment]
Same thing happening here. Being called multiple times in a row..

**Comment by akashbdj** [comment]
I'm also facing the same issue.

Here's the code I've written:

**Parent Component:**
<img width="527" alt="screen shot 2017-12-05 at 8 46 32 am" src="https://user-images.githubusercontent.com/7085282/33588465-ff8e50a2-d998-11e7-8da4-b6e8db9802db.png">


**Options Call**
<img width="533" alt="screen shot 2017-12-05 at 8 45 26 am" src="https://user-images.githubusercontent.com/7085282/33588486-1cc37026-d999-11e7-8663-2918ae78fd0b.png">


**Output :**
<img width="398" alt="screen shot 2017-12-05 at 8 44 48 am" src="https://user-images.githubusercontent.com/7085282/33588517-41bc0d8e-d999-11e7-9e64-0ce7a07234e0.png">

**Comment by parthgandhi7** [comment]
+1.

**Comment by parthgandhi7** [comment]
Aah, my bad , i had specified the `pollInterval=10000` in the config.options passed to graphql function as 2nd parameter. Removing that everything works fine :+1:

**Comment by ahummel25** [comment]
I'm having a similar issue on page load. GraphQL query or mutation firing on page load. 

`Login` function only gets called on `onClick` but query is still firing prematurely and throwing console errors.

```
login(event) {
    event.preventDefault();
    this.props.mutate({ 
      variables: { userName: this.state.userName, password: this.state.password }
    }).then(({ data }) => {
      console.log(data);
    }).catch((error) => {
      console.log(error);
    })
  };`

export default graphql(loginQuery)(withStyles(s)(Login));

**Comment by vivex** [comment]
any update?

**Comment by karthikpamidimarri** [comment]
any updates I am also facing the same issue.

**Comment by newoga** [comment]
To anyone that is having this problem:

It seems like this issue was fixed in `v2` of `react-apollo` 🎉. Many of the internals of `react-apollo` (and its dependencies) were refactored and/or rewritten in `v2` and it appears most of the current development/activity is focused there. I rebased my original reproduction of this issue in [this commit](https://github.com/newoga/react-apollo-issue-660/tree/fixed) and it seems like the behavior is now gone.

If you are running `v1` of `react-apollo` and are having this issue, I recommend updating to `v2` and seeing if that fixes your issue. The migration guide to `v2` is [here](https://www.apollographql.com/docs/react/2.0-migration.html). I've been running `v2` in production for a while now and haven't had any problems.

If you are running `v2` and are having this issue, I recommend cloning the latest version of the [react-apollo-error-template](https://github.com/apollographql/react-apollo-error-template) and trying to create a reproduction there. (I was unsuccessful in doing so)

@jbaxleyiii @helfer @stubailo  I briefly looked at the source code in `v2` and I can't see a scenario where this behavior would still be present. If there aren't any intentions of patching `v1`, I think we can close this issue.
